protocol/messenger_activity_center.go

```diff
@@ -318,11 +399,130 @@ func (m *Messenger) handleActivityCenterAccepted(state *ReceivedMessageState, me
 }
 
 func (m *Messenger) handleActivityCenterDismissed(state *ReceivedMessageState, message protobuf.SyncActivityCenterDismissed) error {
-	resp, err := m.DismissActivityCenterNotifications(context.TODO(), toHexBytes(message.Ids), false)
+	resp, err := m.DismissActivityCenterNotifications(context.TODO(), toHexBytes(message.Ids), message.Clock, false)
 
 	if err != nil {
 		return err
 	}
 
 	return state.Response.Merge(resp)
 }
+
+func (m *Messenger) handleSyncActivityCenterNotificationState(state *ReceivedMessageState, a *protobuf.SyncActivityCenterNotificationState) error {
+	s := &ActivityCenterState{
+		HasSeen:   a.HasSeen,
+		UpdatedAt: a.UpdatedAt,
+	}
+	n, err := m.persistence.UpdateActivityCenterNotificationState(s)
+	if err != nil {
+		return err
+	}
+	if n > 0 {
+		state.Response.SetActivityCenterState(s)
+	}
+	return nil
+}
+
+func (m *Messenger) handleSyncActivityCenterNotifications(state *ReceivedMessageState, a *protobuf.SyncActivityCenterNotifications) error {
+	var notifications []*ActivityCenterNotification
+	for _, n := range a.ActivityCenterNotifications {
+		notification, err := convertActivityCenterNotificationFromProtobuf(n)
+		if err != nil {
+			return err
+		}
+		affectedNum, err := m.persistence.SaveActivityCenterNotification(notification, false)
+		if err != nil {
+			return err
+		}
+		if affectedNum > 0 {
+			notifications = append(notifications, notification)
+		}
+	}
+	response, err := m.processActivityCenterNotifications(notifications, true)
+	if err != nil {
+		return err
+	}
+	return state.Response.Merge(response)
+}
+
+func convertActivityCenterNotificationToProtobuf(n *ActivityCenterNotification) (p *protobuf.SyncActivityCenterNotification, err error) {
+	if n == nil {
+		return nil, errors.New("convertActivityCenterNotificationToProtobuf, n is nil")
+	}
+
+	var (
+		message      []byte
+		replyMessage []byte
+	)
+	if n.Message != nil {
+		message, err = json.Marshal(n.Message)
+		if err != nil {
+			return
+		}
+	}
+	if n.ReplyMessage != nil {
+		replyMessage, err = json.Marshal(n.ReplyMessage)
+		if err != nil {
+			return
+		}
+	}
+	p = &protobuf.SyncActivityCenterNotification{
```

Comment:
Not a huge problem here, but I think it's a bit better to give more expressive names to named parameters. First of all because of the scope of the code is not short (as recommended [here](https://google.github.io/styleguide/go/decisions#variable-names)).

The other reason is that it's relatively easy to shadow single letter variables, and for a named return this can cause a silent bug in the system.

Again, not much of a problem here, just something I try to follow as a guide.